lib/srv/db/sqlserver/kinit/kinit.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/gravitational/teleport/lib/auth/windows"
+)
+
+const (
+	// krb5ConfigEnv sets the location from which kinit will attempt to read a configuration value
+	krb5ConfigEnv = "KRB5_CONFIG"
+)
+
+// ProviderI is a provider interface for a kinit context
+type ProviderI interface {
+	// CreateOrAppendCredentialsCache will create or append to an existing credentials cache
+	CreateOrAppendCredentialsCache(context.Context) error
+	// CacheName is the cache name of this cache
+	CacheName() string
+}
+
+// CommandLineKInit is a command line provider for a kinit context
+type CommandLineKInit struct {
+	authClient auth.ClientI
+
+	userName  string
+	cacheName string
+
+	realmName string
+
+	kdcHostName     string
+	adminServerName string
+
+	certPath string
+	keyPath  string
+
+	ldapCertificate *x509.Certificate
+
+	log logrus.FieldLogger
+}
+
+// NewCommandProvider will return a new command line kinit provider
+func NewCommandProvider(authClient auth.ClientI, user, realm, kdcHost, adminServer string, ldapCA *x509.Certificate) *CommandLineKInit {
+	return &CommandLineKInit{
+		authClient:      authClient,
+		userName:        user,
+		cacheName:       fmt.Sprintf("%s@%s", user, realm),
+		realmName:       realm,
+		kdcHostName:     kdcHost,
+		adminServerName: adminServer,
+
+		ldapCertificate: ldapCA,
+
+		certPath: fmt.Sprintf("%s.pem", user),
+		keyPath:  fmt.Sprintf("%s-key.pem", user),
+		log:      logrus.StandardLogger(),
+	}
+}
+
+// CreateOrAppendCredentialsCache creates or appends to an existing credentials cache.
+func (k *CommandLineKInit) CreateOrAppendCredentialsCache(ctx context.Context) error {
+
+	tmp := filepath.Join(os.TempDir(), "kinit")
+	err := os.MkdirAll(tmp, 0664)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	certPath := filepath.Join(tmp, fmt.Sprintf("%s.pem", k.userName))
+	keyPath := filepath.Join(tmp, fmt.Sprintf("%s-key.pem", k.userName))
+	userCAPath := filepath.Join(tmp, "userca.pem")
+
+	clusterName, err := k.authClient.GetClusterName()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	certPEM, keyPEM, err := windows.CertKeyPEM(ctx, k.userName, k.realmName, time.Second*60*60, clusterName.GetClusterName(), windows.LDAPConfig{
+		Addr:               k.kdcHostName,
+		Domain:             k.realmName,
+		Username:           k.userName,
+		InsecureSkipVerify: false,
+		ServerName:         k.adminServerName,
+		CA:                 k.ldapCertificate,
+	}, k.authClient)
+
+	userCA, err := k.authClient.GetCertAuthority(ctx, types.CertAuthID{
+		Type:       types.UserCA,
+		DomainName: clusterName.GetClusterName(),
+	}, true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// get the user CA certificate bytes
+	var caCert []byte
+	keyPairs := userCA.GetActiveKeys().TLS
+	for _, keyPair := range keyPairs {
+		if keyPair.KeyType == types.PrivateKeyType_RAW {
+			caCert = keyPair.Cert
+		}
+	}
+
+	if caCert == nil {
+		return trace.Wrap(errors.New("no certificate authority was found in userCA active keys"))
```

Comment:
We should probably update the trace package to make this easier. Maybe add something like `trace.New()`..